### PR TITLE
fix: [DEL-4328]: Make threadpool for log streaming task client as static

### PR DESCRIPTION
**Pull request by gaurav-jain1**
## Describe your changes
As part of this patchset, made change to make threadpool static, so that we don't end up initialising on every object creation.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37392)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37392)

File: 260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java [hunk 1]
```diff
@@ -64,10 +65,14 @@ public class LogStreamingTaskClient implements ILogStreamingTaskClient {
   private final String token;
   private final String accountId;
   private final String baseLogKey;
-  private ScheduledExecutorService scheduledExecutorService;
+  private static ScheduledExecutorService scheduledExecutorService = new ScheduledThreadPoolExecutor(50,
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java:68]: making it static makes complete sense.
I have two questions, may not be super valid points
1. why not create a provider to provide the thread pool executor ? so all the logStreamTaskClient will use the same pool
2. why 50 threads ?

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 75 at r2](https://reviewable.io/reviews/harness/harness-core/37392#-NBnaKsQB9y0guM8Io2F:-NBnaKsQB9y0guM8Io2G:buc3kym) ([raw file](https://github.com/harness/harness-core/blob/5f8d6f8b66fa4a7fd44ece018d50ea1dc7c82863/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L75)):*
> ```Java
>   @Deprecated private final String appId;
>   @Deprecated private final String activityId;
>   ScheduledFuture scheduledFuture;
> ```

private
___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 118 at r2](https://reviewable.io/reviews/harness/harness-core/37392#-NBna0X-6OavEnbthR8v:-NBna0X-6OavEnbthR8w:b-28lonn) ([raw file](https://github.com/harness/harness-core/blob/5f8d6f8b66fa4a7fd44ece018d50ea1dc7c82863/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L118)):*
> ```Java
>     } finally {
>       if (scheduledFuture != null) {
>         scheduledFuture.cancel(true);
> ```

I think we should attempt to wait (with a timeout) for the future here instead. Otherwise you might loose the last batch of logs.
___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 120 at r2](https://reviewable.io/reviews/harness/harness-core/37392#-NBn_k3KAMpzmkYE77Mu:-NBn_k3KAMpzmkYE77Mv:b-8r64er) ([raw file](https://github.com/harness/harness-core/blob/5f8d6f8b66fa4a7fd44ece018d50ea1dc7c82863/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L120)):*
> ```Java
>         scheduledFuture.cancel(true);
>       }
>       scheduledExecutorService.shutdownNow();
> ```

You should not shutdown on closeStream since the service is static now. It should be on some other event (i.e. delegate shutdown, but possibly no need to deal with it for now)

**Review by xingchi-jin — COMMENTED** [review]
LG after setting the pool size min-max, and fix the shutdown call. 

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392#-:-NBpYOMSDo__zUNDqKZU:b-nyg8a0)* status: 1 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 120 at r2](https://reviewable.io/reviews/harness/harness-core/37392#-NBn_k3KAMpzmkYE77Mu:-NBpb2no1crAdBzQFoAU:b-88kfwe) ([raw file](https://github.com/harness/harness-core/blob/5f8d6f8b66fa4a7fd44ece018d50ea1dc7c82863/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L120)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Make sense. I think d'tor for this class will take care of garbage collecting this.
</blockquote></details>

+1, if the future tasks of this object has been canceled. Shutdown the "global" executor here looks unnecessary.

**Review by xingchi-jin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392)* status: 2 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/37392#-NBzw91P1h6QQsZBXwCv:-NBzw91P1h6QQsZBXwCw:baxcmu) ([raw file](https://github.com/harness/harness-core/blob/47f29c66a4d66bfbdce49ac6f473dafd389c5edb/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L74)):*
> ```Java
>   private ScheduledFuture scheduledFuture;
>   private final ITaskProgressClient taskProgressClient;
>   private String logKey;
> ```

Looks like there are more than 1 logKey in the map, as every time "writeLogLine", there could create a new logkey in the map. If no such a logKey that is special enough to be the member of this class, we should remove it from here.
But, correct me if i m wrong.

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392)* status: 2 file(s) to review. 2 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 118 at r2](https://reviewable.io/reviews/harness/harness-core/37392#-NBna0X-6OavEnbthR8v:-NC3hD_E1XR_49_f2NQK:bd8ouc9) ([raw file](https://github.com/harness/harness-core/blob/5f8d6f8b66fa4a7fd44ece018d50ea1dc7c82863/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L118)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

> mpt to wait (with a timeout) for the future here instead. Otherwise you might loose the last batch of logs.
We don't need to wait here, since we already have waited above in this method. By the time control is here, we should end up submitting all logs.

</blockquote></details>

Done, since now we are executing closeStream asynchronously, made this argument to false, so that we can allow existing tasks to run.
___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/37392#-NBzw91P1h6QQsZBXwCv:-NC3gfKw9EX6Fw5tbbNy:bxovjis) ([raw file](https://github.com/harness/harness-core/blob/47f29c66a4d66bfbdce49ac6f473dafd389c5edb/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L74)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

Looks like there are more than 1 logKey in the map, as every time "writeLogLine", there could create a new logkey in the map. If no such a logKey that is special enough to be the member of this class, we should remove it from here.
But, correct me if i m wrong.
</blockquote></details>

Yes, actually now I felt we shouldn't change logkey there as we don't know what can consequences. I am putting log.error to identify when logkey doesn't matches, and later we can clean this code further. But for now I am using the logkey being passed as argument.

**Review by xingchi-jin — DISMISSED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392#-:-NCE1CRh3-gUxlPAgy1f:bnfp4nl)* status: 2 file(s) to review. 1 change requests. 3. unresolved discussions. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/37392#-NBzw91P1h6QQsZBXwCv:-NCE0zl25kpb2ian3vSg:bxyao0j) ([raw file](https://github.com/harness/harness-core/blob/47f29c66a4d66bfbdce49ac6f473dafd389c5edb/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L74)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Yes, actually now I felt we shouldn't change logkey there as we don't know what can consequences. I am putting log.error to identify when logkey doesn't matches, and later we can clean this code further. But for now I am using the logkey being passed as argument.
</blockquote></details>

make sense. also, if this private member "logKey" is no longer used, it can be removed. Otherwise, the change looks good to me, check if others have comments

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37392)* status: 1 change requests. 3. unresolved discussions.

**Pull request merged**